File: examples/gallery.py
import asyncio
import os

from loguru import logger

import awr

# 登录账号
account = os.environ.get("AWR_ACCOUNT")
if not account:
    logger.error("AWR_ACCOUNT not set")
    exit(1)
account = int(account)

password = os.environ.get("AWR_PASSWORD")
if not password:
    logger.error("AWR_PASSWORD not set")
    exit(1)

# 测试用好友账号
test_account = os.environ.get("QQ_TEST_ACCOUNT")

async def main():
    client = await awr.Dynamic().login(int(account), "./bots")

    uin = client.uin
    logger.info(f"账号: {uin}")

    is_online = client.is_online()
    logger.info(f"在线状态: {is_online}")

    info = await client.get_account_info()
    logger.info(f"昵称: {info.nickname}")
    logger.info(f"年龄: {info.age}")
    logger.info(f"性别: {info.gender}")

    logger.info("################################")

    friend_list = await client.get_friend_list()
    logger.info(f"好友数量: {friend_list.total_count}")
    logger.info(f"在线好友数量: {friend_list.online_count}")
    friend_count = len(list(friend_list.friends()))
    logger.info(f"遍历好友列表, 总数: {friend_count}")
    logger.info(
        f"好友分组: {[friend_group.name for friend_group in friend_list.friend_groups()]}"
    )

    logger.info("################################")

    if test_account and (friend := friend_list.find_friend(int(test_account))):
        logger.info(f"好友账号: {friend.uin}")
        logger.info(f"好友昵称: {friend.nickname}")
        logger.info(f"好友备注: {friend.remark}")
        logger.info(f"好友分组: {friend.group_id}")

        selector = friend.as_selector()
        logger.info("发送戳一戳")
        await selector.poke()
        await asyncio.sleep(0.5)

        logger.info("发送消息")
        receipt = await client.friend(1779724477).send("在吗", awr.Face(0))
        await asyncio.sleep(2)
        logger.info("撤回消息")
        await receipt.recall()

        logger.info("构造消息链")
        message = await awr.MessageContent.build_friend_message("您好")
        await friend.send(message)

    logger.info("################################")

    group_list = await client.get_group_list()
    logger.info(f"群数量: {len(group_list)}")
    logger.info(f"群列表: {[group.name for group in group_list]}")

    logger.info("测试完毕")
    await client.alive()


try:
    asyncio.run(main())
except KeyboardInterrupt:
    import sys

    sys.exit(0)

File: examples/gallery.rs
#![feature(let_chains)]
//! libawr 的示例代码。

use anyhow::Result;
use libawr::{login, msg};
use ricq::msg::elem::Face;
use std::time::Duration;

#[tokio::main]
async fn main() -> Result<()> {
    tracing_subscriber::fmt::init();

    let account = std::env::var("AWR_ACCOUNT")?.parse()?;
    let password = std::env::var("AWR_PASSWORD")?;
    let test_account = std::env::var("AWR_TEST_ACCOUNT")
        .ok()
        .map(|s| s.parse())
        .transpose()?;

    let (client, alive) = login!(
        account,
        password = &password,
        protocol = ricq::Protocol::IPad,
    )
    .await?;

    tracing::info!("登录成功！");

    tracing::info!("账号: {}", client.uin);
    tracing::info!("在线状态: {}", client.is_online());

    tracing::info!("################################");

    let friend_list = client.get_friend_list().await?;
    tracing::info!("好友数量: {}", friend_list.total_count);
    tracing::info!("在线好友数量: {}", friend_list.online_count);
    let friend_count = friend_list.friends().len();
    tracing::info!("好友总数: {}", friend_count);
    tracing::info!(
        "好友分组: {}",
        friend_list
            .friend_groups()
            .values()
            .map(|group| group.name.as_str())
            .collect::<Vec<_>>()
            .join(", ")
    );

    tracing::info!("################################");

    if let Some(test_account) = test_account
        && let Some(friend) = client.get_friend(test_account).await? {
        tracing::debug!("好友信息: {:?}", friend);
        tracing::info!("好友账号: {}", friend.uin);
        tracing::info!("好友账号昵称: {}", friend.nickname);
        tracing::info!("好友账号备注: {}", friend.remark);
        tracing::info!("好友账号分组: {}", friend.group_id);

        tracing::info!("发送好友戳一戳");
        friend.poke().await?;

        tracing::info!("发送好友消息");
        let receipt = friend.send(msg!("在吗", Face::new(0))).await?;
        tokio::time::sleep(Duration::from_secs(2)).await;
        tracing::info!("撤回好友消息");
        receipt.recall().await?;
    }

    tokio::select! {
        _ = count_down(5) => {
            tracing::info!("等待结束，自动退出");
        },
        _ = alive.auto_reconnect() => {
            tracing::error!("自动重连失败");
        }
    }

    Ok(())
}

async fn count_down(max: u64) {
    for i in 0..max {
        tracing::info!("保持连接中，剩余: {}", max - i);
        tokio::time::sleep(Duration::from_secs(1)).await;
    }
}
